File: code/cpp/ch07/practice/practice_01_all_divisors.cpp
/*
 * Practice 1: All Divisors
 * ========================
 * Chapter 7: Number Wizardry -- Math for Programmers
 *
 * PROBLEM:
 *   Given a positive integer n, return ALL its divisors in sorted order.
 *
 * EXAMPLES:
 *   solve(36) -> [1, 2, 3, 4, 6, 9, 12, 18, 36]
 *   solve(1)  -> [1]
 *   solve(7)  -> [1, 7]
 *   solve(12) -> [1, 2, 3, 4, 6, 12]
 *
 * CONSTRAINTS:
 *   1 <= n <= 10^9
 *
 * INSTRUCTIONS:
 *   Replace the body with your solution.
 *   Hint: You only need to check divisors up to sqrt(n).
 *   If i divides n, then both i and n/i are divisors.
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

vector<int> solve(int n) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> result = solve(n);
    for (int i = 0; i < (int)result.size(); i++) {
        if (i > 0) cout << " ";
        cout << result[i];
    }
    cout << endl;
    return 0;
}

File: code/java/ch07/practice/Practice01AllDivisors.java
package ch07.practice;

import java.util.*;

/**
 * Practice 01: All Divisors
 * ==============================
 * Chapter 7: Number Wizardry
 *
 * PROBLEM: Given a positive integer n, return a sorted list of ALL
 *          its divisors.
 *
 * EXAMPLES:
 *   solve(36) = [1, 2, 3, 4, 6, 9, 12, 18, 36]
 *   solve(1)  = [1]
 *   solve(7)  = [1, 7]
 *   solve(12) = [1, 2, 3, 4, 6, 12]
 *
 * CONSTRAINTS:
 *   1 <= n <= 10^9
 *
 * HINT: You only need to check up to sqrt(n). If i divides n,
 *       then n/i also divides n.
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Practice01AllDivisors {
    public static List<Integer> solve(int n) {
        // TODO: Replace this with your solution
        return new ArrayList<>();
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(n));
        sc.close();
    }
}
